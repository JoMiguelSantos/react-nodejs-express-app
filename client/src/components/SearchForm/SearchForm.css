.search-form {
  font-size: 1rem;
}

.search-form > label {
  display: flex;
  align-items: center;
  margin: 1rem;
  padding-right: 0.5rem;
}

label input {
  /* margin: 0.5rem; */
  padding-left: 0.5rem;
  width: 80%;
  margin-left: auto;
}

.btn-search {
  margin: 0 2rem;
  padding: 0.2rem 1rem;
  width: 10rem;
  font-size: 1rem;
}

.btn-search:hover {
  background-color: azure;
  cursor: pointer;
}
