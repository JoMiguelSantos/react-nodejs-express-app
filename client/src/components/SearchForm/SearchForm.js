import React, { useState } from "react";
import { bindActionCreators } from "redux";
import { useDispatch } from "react-redux";
import { withRouter } from "react-router-dom";
import { fetchRepos } from "../../store/actions";
import "./SearchForm.css";

const SearchForm = ({ history }) => {
  const [isEmpty, setEmpty] = useState(false);
  const dispatch = useDispatch();
  const boundFetchRepos = bindActionCreators(fetchRepos, dispatch);

  const handleSubmit = async e => {
    e.preventDefault();

    let searchTerms = ``;
    [...e.target.elements].forEach((el, index) => {
      if (el.name && el.value) {
        if (index === 0) {
          searchTerms = searchTerms.concat(`${el.name}=${el.value}`);
        } else {
          searchTerms = searchTerms.concat(`&${el.name}=${el.value}`);
        }
      }
    });
    console.log(!searchTerms);

    setEmpty(!searchTerms);
    const repos = searchTerms ? await boundFetchRepos(searchTerms) : "";
    console.log(repos);

    if (!repos) return;
    if (repos && repos.length > 0) history.push("/repos");
  };

  return (
    <form className="search-form" onSubmit={handleSubmit}>
      <label>
        Name
        <input
          name="name"
          type="text"
          placeholder="e.g. node-express-app"
        ></input>
      </label>
      <label>
        Description
        <input
          name="description"
          type="text"
          placeholder="e.g. A node express app built from scratch"
        ></input>
      </label>
      <label>
        README
        <input
          name="readme"
          type="text"
          placeholder="e.g. API documentation on how to use it"
        ></input>
      </label>
      <label>
        Main Language
        <input
          name="language"
          type="text"
          placeholder="e.g. JavaScript"
        ></input>
      </label>
      <label>
        Topic
        <input
          name="topic"
          type="text"
          placeholder="e.g. Server Side Rendering"
        ></input>
      </label>
      <button className="btn-search">Search</button>
      {isEmpty && (
        <p className="empty-submit">
          Please fill in at least one of the fields
        </p>
      )}
    </form>
  );
};

export default withRouter(SearchForm);
