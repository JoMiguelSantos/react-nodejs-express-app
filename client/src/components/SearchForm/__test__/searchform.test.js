import SearchForm from "../SearchForm";
import React from "react";
import { createStore } from "redux";
import { Provider } from "react-redux";
import { render, fireEvent } from "@testing-library/react";
import reducer, { initialState } from "../../../store/reducers/repos";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";

function withRouterRedux({
  initialState,
  store = createStore(reducer, initialState)
} = {}) {
  const history = createMemoryHistory();
  return {
    ...render(
      <Router history={history}>
        <Provider store={store}>
          <SearchForm />
        </Provider>
      </Router>
    ),
    store,
    history
  };
}

describe("<SearchForm/> unit tests", () => {
  // test snapshot which should not change often
  it("matches the snapshot", () => {
    const { container } = withRouterRedux();
    expect(container).toMatchSnapshot();
  });

  it("shows validation text when search is clicked without input data", () => {
    const { getByText } = withRouterRedux();
    fireEvent.click(getByText(/search/i));
    expect(
      getByText("Please fill in at least one of the fields")
    ).toBeVisible();
  });
});
