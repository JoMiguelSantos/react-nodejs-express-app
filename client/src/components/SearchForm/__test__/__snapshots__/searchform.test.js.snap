// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchForm/> unit tests matches the snapshot 1`] = `
<div>
  <form
    class="search-form"
  >
    <label>
      Name
      <input
        name="name"
        placeholder="e.g. node-express-app"
        type="text"
      />
    </label>
    <label>
      Description
      <input
        name="description"
        placeholder="e.g. A node express app built from scratch"
        type="text"
      />
    </label>
    <label>
      README
      <input
        name="readme"
        placeholder="e.g. API documentation on how to use it"
        type="text"
      />
    </label>
    <label>
      Main Language
      <input
        name="language"
        placeholder="e.g. JavaScript"
        type="text"
      />
    </label>
    <label>
      Topic
      <input
        name="topic"
        placeholder="e.g. Server Side Rendering"
        type="text"
      />
    </label>
    <button
      class="btn-search"
    >
      Search
    </button>
  </form>
</div>
`;
